File: doa/database/persistence.py
"""Data persistence layer for TradeWizard DOA."""

import logging
import json
from typing import List, Optional, Dict, Any
from datetime import datetime
from uuid import uuid4
from postgrest.exceptions import APIError

from models.types import (
    MarketBriefingDocument,
    AgentSignal,
    TradeRecommendation,
    Thesis,
    DebateRecord,
    ConsensusProbability,
    AuditEntry
)
from database.supabase_client import SupabaseClient
from database.db_types import (
    MarketInsert,
    RecommendationInsert,
    AgentSignalInsert,
    AnalysisHistoryInsert
)
from utils.result import Result, Ok, Err

logger = logging.getLogger(__name__)


class PersistenceLayer:
    """
    Data persistence layer with fallback to in-memory storage.
    
    Implements save/load operations for market data, agent signals,
    and recommendations with graceful degradation on database errors.
    """
    
    def __init__(self, supabase_client: SupabaseClient):
        """
        Initialize persistence layer.
        
        Args:
            supabase_client: Supabase client instance
        """
        self.client = supabase_client
        self._in_memory_markets: Dict[str, MarketBriefingDocument] = {}
        self._in_memory_signals: Dict[str, List[AgentSignal]] = {}
        self._in_memory_recommendations: Dict[str, TradeRecommendation] = {}
        self._fallback_mode = False
    
    async def save_market_data(
        self,
        mbd: MarketBriefingDocument
    ) -> Result[str, str]:
        """
        Save market briefing document to database.
        
        Args:
            mbd: Market briefing document to save
            
        Returns:
            Ok(market_id) on success, Err(message) on failure
        """
        # Check if database is available
        if not self.client.is_connected():
            logger.warning("Database not connected, using in-memory storage")
            self._in_memory_markets[mbd.condition_id] = mbd
            self._fallback_mode = True
            return Ok(mbd.condition_id)
        
        try:
            # Check if market already exists
            existing = self.client.client.table("markets").select("id").eq(
                "condition_id", mbd.condition_id
            ).execute()
            
            if existing.data:
                # Update existing market
                market_id = existing.data[0]["id"]
                update_data = {
                    "question": mbd.question,
                    "description": mbd.event_context.event_description if mbd.event_context else None,
                    "event_type": mbd.event_type,
                    "market_probability": mbd.current_probability,
                    "volume_24h": mbd.volume_24h,
                    "liquidity": mbd.liquidity_score * 100000,  # Convert score to approximate liquidity
                    "updated_at": datetime.now().isoformat()
                }
                
                response = self.client.client.table("markets").update(
                    update_data
                ).eq("id", market_id).execute()
                
                logger.info(f"Updated market data for condition_id: {mbd.condition_id}")
                return Ok(market_id)
            else:
                # Insert new market - use condition_id as the id
                insert_data = {
                    "id": mbd.condition_id,  # Use condition_id as the primary key
                    "condition_id": mbd.condition_id,
                    "question": mbd.question,
                    "event_type": mbd.event_type,
                    "market_probability": mbd.current_probability,
                    "volume_24h": mbd.volume_24h,
                    "liquidity": mbd.liquidity_score * 100000,
                    "status": "active"
                }
                
                # Add optional fields only if they exist
                if mbd.event_context and mbd.event_context.event_description:
                    insert_data["description"] = mbd.event_context.event_description
                
                response = self.client.client.table("markets").insert(
                    insert_data
                ).execute()
                
                market_id = response.data[0]["id"]
                logger.info(f"Saved new market data for condition_id: {mbd.condition_id}")
                return Ok(market_id)
        
        except APIError as e:
            logger.error(f"Database error saving market data: {e}")
            # Fallback to in-memory
            self._in_memory_markets[mbd.condition_id] = mbd
            self._fallback_mode = True
            return Err(f"Database error: {e}")
        
        except Exception as e:
            logger.error(f"Unexpected error saving market data: {e}")
            # Fallback to in-memory
            self._in_memory_markets[mbd.condition_id] = mbd
            self._fallback_mode = True
            return Err(f"Unexpected error: {e}")
    
    async def save_agent_signals(
        self,
        condition_id: str,
        market_id: str,
        recommendation_id: Optional[str],
        signals: List[AgentSignal]
    ) -> Result[None, str]:
        """
        Save agent signals to database.
        
        Args:
            condition_id: Market condition ID (for in-memory fallback)
            market_id: Market UUID from database
            recommendation_id: Recommendation UUID (optional)
            signals: List of agent signals to save
            
        Returns:
            Ok(None) on success, Err(message) on failure
        """
        if not signals:
            return Ok(None)
        
        # Check if database is available
        if not self.client.is_connected():
            logger.warning("Database not connected, using in-memory storage")
            if condition_id not in self._in_memory_signals:
                self._in_memory_signals[condition_id] = []
            self._in_memory_signals[condition_id].extend(signals)
            self._fallback_mode = True
            return Ok(None)
        
        try:
            # Prepare signals for insertion
            signals_data = []
            for signal in signals:
                # Determine agent type from agent name
                agent_type = self._get_agent_type(signal.agent_name)
                
                signal_insert = AgentSignalInsert(
                    market_id=market_id,
                    recommendation_id=recommendation_id,
                    agent_name=signal.agent_name,
                    agent_type=agent_type,
                    fair_probability=signal.fair_probability,
                    confidence=signal.confidence,
                    direction=signal.direction,
                    key_drivers={"drivers": signal.key_drivers, "risks": signal.risk_factors},
                    metadata=signal.metadata
                )
                signals_data.append(signal_insert.model_dump(exclude_none=True, mode='json'))
            
            # Insert all signals
            response = self.client.client.table("agent_signals").insert(
                signals_data
            ).execute()
            
            logger.info(f"Saved {len(signals)} agent signals for market_id: {market_id}")
            return Ok(None)
        
        except APIError as e:
            logger.error(f"Database error saving agent signals: {e}")
            # Fallback to in-memory
            if condition_id not in self._in_memory_signals:
                self._in_memory_signals[condition_id] = []
            self._in_memory_signals[condition_id].extend(signals)
            self._fallback_mode = True
            return Err(f"Database error: {e}")
        
        except Exception as e:
            logger.error(f"Unexpected error saving agent signals: {e}")
            # Fallback to in-memory
            if condition_id not in self._in_memory_signals:
                self._in_memory_signals[condition_id] = []
            self._in_memory_signals[condition_id].extend(signals)
            self._fallback_mode = True
            return Err(f"Unexpected error: {e}")
    
    def _get_agent_type(self, agent_name: str) -> str:
        """Determine agent type from agent name."""
        agent_type_map = {
            "market_microstructure": "mvp",
            "probability_baseline": "mvp",
            "risk_assessment": "mvp",
            "breaking_news": "event_intelligence",
            "event_impact": "event_intelligence",
            "polling_intelligence": "polling_statistical",
            "historical_pattern": "polling_statistical",
            "media_sentiment": "sentiment_narrative",
            "social_sentiment": "sentiment_narrative",
            "narrative_velocity": "sentiment_narrative",
            "momentum": "price_action",
            "mean_reversion": "price_action",
            "catalyst": "event_scenario",
            "tail_risk": "event_scenario"
        }
        return agent_type_map.get(agent_name, "other")
    
    async def save_recommendation(
        self,
        recommendation: TradeRecommendation,
        market_id: str
    ) -> Result[str, str]:
        """
        Save trade recommendation to database.
        
        Args:
            recommendation: Trade recommendation to save
            market_id: Market UUID from database
            
        Returns:
            Ok(recommendation_id) on success, Err(message) on failure
        """
        # Check if database is available
        if not self.client.is_connected():
            logger.warning("Database not connected, using in-memory storage")
            self._in_memory_recommendations[recommendation.condition_id] = recommendation
            self._fallback_mode = True
            return Ok(recommendation.condition_id)
        
        try:
            # Map action to direction
            direction_map = {
                "LONG_YES": "LONG_YES",
                "LONG_NO": "LONG_NO",
                "NO_TRADE": "NO_TRADE"
            }
            
            # Map regime to confidence
            confidence_map = {
                "high-confidence": "high",
                "moderate-confidence": "moderate",
                "high-uncertainty": "low"
            }
            
            # Prepare recommendation for insertion
            rec_insert = RecommendationInsert(
                market_id=market_id,
                direction=direction_map.get(recommendation.action, recommendation.action),
                fair_probability=recommendation.metadata.consensus_probability,
                market_edge=recommendation.metadata.edge,
                expected_value=recommendation.expected_value,
                confidence=confidence_map.get(recommendation.metadata.regime, "moderate"),
                entry_zone_min=recommendation.entry_zone[0],
                entry_zone_max=recommendation.entry_zone[1],
                target_zone_min=recommendation.target_zone[0],
                target_zone_max=recommendation.target_zone[1],
                explanation=recommendation.explanation.summary,
                core_thesis=recommendation.explanation.core_thesis,
                catalysts={"catalysts": recommendation.explanation.key_catalysts},
                risks={"scenarios": recommendation.explanation.failure_scenarios}
            )
            
            # Insert recommendation
            response = self.client.client.table("recommendations").insert(
                rec_insert.model_dump(exclude_none=True, mode='json')
            ).execute()
            
            recommendation_id = response.data[0]["id"]
            logger.info(f"Saved recommendation for condition_id: {recommendation.condition_id}")
            return Ok(recommendation_id)
        
        except APIError as e:
            logger.error(f"Database error saving recommendation: {e}")
            # Fallback to in-memory
            self._in_memory_recommendations[recommendation.condition_id] = recommendation
            self._fallback_mode = True
            return Err(f"Database error: {e}")
        
        except Exception as e:
            logger.error(f"Unexpected error saving recommendation: {e}")
            # Fallback to in-memory
            self._in_memory_recommendations[recommendation.condition_id] = recommendation
            self._fallback_mode = True
            return Err(f"Unexpected error: {e}")

    async def save_analysis_history(
        self,
        condition_id: str,
        market_id: str,
        analysis_timestamp: int,
        agent_count: int,
        consensus: ConsensusProbability,
        recommendation_action: str,
        duration_ms: int,
        cost_usd: float,
        status: str = "success",
        error_message: Optional[str] = None
    ) -> Result[None, str]:
        """
        Save complete analysis history to database.
        
        Args:
            condition_id: Market condition ID (for in-memory fallback)
            market_id: Market UUID from database
            analysis_timestamp: Unix timestamp of analysis
            agent_count: Number of agents that participated
            consensus: Consensus probability result
            recommendation_action: Final recommendation action
            duration_ms: Analysis duration in milliseconds
            cost_usd: Estimated cost in USD
            status: Analysis status (success, failed, partial)
            error_message: Error message if failed
            
        Returns:
            Ok(None) on success, Err(message) on failure
        """
        # Check if database is available
        if not self.client.is_connected():
            logger.warning("Database not connected, skipping analysis history")
            self._fallback_mode = True
            return Ok(None)
        
        try:
            # Prepare history data
            history_insert = AnalysisHistoryInsert(
                market_id=market_id,
                analysis_type="initial",  # Could be parameterized
                status=status,
                duration_ms=duration_ms,
                cost_usd=cost_usd,
                agents_used={
                    "agent_count": agent_count,
                    "consensus_probability": consensus.consensus_probability,
                    "disagreement_index": consensus.disagreement_index,
                    "regime": consensus.regime,
                    "recommendation_action": recommendation_action
                },
                error_message=error_message
            )
            
            # Insert history
            response = self.client.client.table("analysis_history").insert(
                history_insert.model_dump(exclude_none=True, mode='json')
            ).execute()
            
            logger.info(f"Saved analysis history for condition_id: {condition_id}")
            return Ok(None)
        
        except APIError as e:
            logger.error(f"Database error saving analysis history: {e}")
            return Err(f"Database error: {e}")
        
        except Exception as e:
            logger.error(f"Unexpected error saving analysis history: {e}")
            return Err(f"Unexpected error: {e}")
    
    async def get_historical_signals(
        self,
        condition_id: str,
        agent_name: str,
        limit: int = 3
    ) -> Result[List[AgentSignal], str]:
        """
        Retrieve historical agent signals for memory context.
        
        Args:
            condition_id: Market condition ID
            agent_name: Name of the agent
            limit: Maximum number of signals to retrieve
            
        Returns:
            Ok(signals) on success, Err(message) on failure
        """
        # Check in-memory first if in fallback mode
        if self._fallback_mode:
            logger.info("Using in-memory storage for historical signals")
            if condition_id in self._in_memory_signals:
                signals = [
                    s for s in self._in_memory_signals[condition_id]
                    if s.agent_name == agent_name
                ]
                return Ok(signals[-limit:] if len(signals) > limit else signals)
            return Ok([])
        
        # Check if database is available
        if not self.client.is_connected():
            logger.warning("Database not connected, no historical signals available")
            return Ok([])
        
        try:
            # First, get the market_id from condition_id
            market_response = self.client.client.table("markets").select("id").eq(
                "condition_id", condition_id
            ).execute()
            
            if not market_response.data:
                # No market found, return empty list
                logger.info(f"No market found for condition_id: {condition_id}")
                return Ok([])
            
            market_id = market_response.data[0]["id"]
            
            # Query historical signals using market_id
            response = self.client.client.table("agent_signals").select("*").eq(
                "market_id", market_id
            ).eq(
                "agent_name", agent_name
            ).order(
                "created_at", desc=True
            ).limit(limit).execute()
            
            # Convert to AgentSignal objects
            signals = []
            for row in response.data:
                # Extract key_drivers and risk_factors from key_drivers JSON field
                key_drivers_data = row.get("key_drivers", {})
                if isinstance(key_drivers_data, dict):
                    key_drivers = key_drivers_data.get("drivers", [])
                    risk_factors = key_drivers_data.get("risks", [])
                else:
                    key_drivers = []
                    risk_factors = []
                
                signal = AgentSignal(
                    agent_name=row["agent_name"],
                    timestamp=int(row["created_at"].timestamp()) if isinstance(row["created_at"], datetime) else int(datetime.fromisoformat(row["created_at"].replace('Z', '+00:00')).timestamp()),
                    confidence=row["confidence"],
                    direction=row["direction"],
                    fair_probability=row["fair_probability"],
                    key_drivers=key_drivers,
                    risk_factors=risk_factors,
                    metadata=row.get("metadata", {})
                )
                signals.append(signal)
            
            logger.info(f"Retrieved {len(signals)} historical signals for {agent_name}")
            return Ok(signals)
        
        except APIError as e:
            logger.error(f"Database error retrieving historical signals: {e}")
            return Err(f"Database error: {e}")
        
        except Exception as e:
            logger.error(f"Unexpected error retrieving historical signals: {e}")
            return Err(f"Unexpected error: {e}")
    
    async def get_market_data(
        self,
        condition_id: str
    ) -> Result[Optional[MarketBriefingDocument], str]:
        """
        Retrieve market briefing document from database.
        
        Args:
            condition_id: Market condition ID
            
        Returns:
            Ok(mbd) on success, Err(message) on failure
        """
        # Check in-memory first if in fallback mode
        if self._fallback_mode:
            logger.info("Using in-memory storage for market data")
            mbd = self._in_memory_markets.get(condition_id)
            return Ok(mbd)
        
        # Check if database is available
        if not self.client.is_connected():
            logger.warning("Database not connected")
            return Ok(None)
        
        try:
            # Query market data
            response = self.client.client.table("markets").select("*").eq(
                "condition_id", condition_id
            ).execute()
            
            if not response.data:
                return Ok(None)
            
            row = response.data[0]
            
            # Reconstruct MarketBriefingDocument
            from models.types import EventContext, StreamlinedEventMetadata
            
            event_context = None
            if row.get("event_context"):
                event_context = EventContext(**row["event_context"])
            
            metadata = StreamlinedEventMetadata(**row["metadata"])
            
            mbd = MarketBriefingDocument(
                market_id=row["market_id"],
                condition_id=row["condition_id"],
                event_type=row["event_type"],
                question=row["question"],
                resolution_criteria=row["resolution_criteria"],
                expiry_timestamp=row["expiry_timestamp"],
                current_probability=row["current_probability"],
                liquidity_score=row["liquidity_score"],
                bid_ask_spread=row["bid_ask_spread"],
                volatility_regime=row["volatility_regime"],
                volume_24h=row["volume_24h"],
                event_context=event_context,
                keywords=row.get("keywords"),
                metadata=metadata
            )
            
            logger.info(f"Retrieved market data for condition_id: {condition_id}")
            return Ok(mbd)
        
        except APIError as e:
            logger.error(f"Database error retrieving market data: {e}")
            return Err(f"Database error: {e}")
        
        except Exception as e:
            logger.error(f"Unexpected error retrieving market data: {e}")
            return Err(f"Unexpected error: {e}")
    
    def is_fallback_mode(self) -> bool:
        """
        Check if persistence layer is in fallback mode.
        
        Returns:
            True if using in-memory storage, False if using database
        """
        return self._fallback_mode
    
    def clear_in_memory_cache(self) -> None:
        """Clear all in-memory cached data."""
        self._in_memory_markets.clear()
        self._in_memory_signals.clear()
        self._in_memory_recommendations.clear()
        logger.info("Cleared in-memory cache")

File: tradewizard-agents/src/database/persistence.ts
/**
 * Database Persistence Layer
 *
 * This module provides a high-level interface for storing and retrieving
 * market analysis data from Supabase PostgreSQL.
 */

import type { SupabaseClientManager } from './supabase-client.js';
import type { TablesInsert, TablesUpdate, Json } from './types.js';
import type { TradeRecommendation, AgentSignal, MarketId } from '../models/types.js';
import { retryDatabaseOperation } from '../utils/retry-logic.js';

/**
 * Market data for database storage
 */
export interface MarketData {
  conditionId: string;
  question: string;
  description?: string;
  eventType: string;
  marketProbability?: number;
  volume24h?: number;
  liquidity?: number;
  status?: 'active' | 'inactive' | 'resolved';
  resolvedOutcome?: string;
  trendingScore?: number;
}

/**
 * Analysis record for history tracking
 */
export interface AnalysisRecord {
  type: 'initial' | 'update' | 'manual';
  status: 'success' | 'failed' | 'partial';
  durationMs?: number;
  costUsd?: number;
  agentsUsed?: string[];
  errorMessage?: string;
}

/**
 * Database Persistence Interface
 */
export interface DatabasePersistence {
  /**
   * Store or update a market
   * @returns market_id (TEXT - can be string or number converted to string)
   */
  upsertMarket(market: MarketData): Promise<string>;

  /**
   * Store a recommendation
   * @returns recommendation_id (UUID)
   */
  storeRecommendation(marketId: MarketId, recommendation: TradeRecommendation): Promise<string>;

  /**
   * Store agent signals
   */
  storeAgentSignals(
    marketId: MarketId,
    recommendationId: string,
    signals: AgentSignal[]
  ): Promise<void>;

  /**
   * Record analysis history
   */
  recordAnalysis(marketId: MarketId, analysis: AnalysisRecord): Promise<void>;

  /**
   * Get markets needing update
   * @param updateIntervalMs - Minimum time since last analysis in milliseconds
   */
  getMarketsForUpdate(updateIntervalMs: number): Promise<MarketData[]>;

  /**
   * Mark market as resolved
   */
  markMarketResolved(marketId: MarketId, outcome: string): Promise<void>;

  /**
   * Get latest recommendation for a market
   */
  getLatestRecommendation(marketId: MarketId): Promise<TradeRecommendation | null>;
}

/**
 * Database Persistence Implementation
 */
export class DatabasePersistenceImpl implements DatabasePersistence {
  constructor(private clientManager: SupabaseClientManager) {}

  /**
   * Normalize market ID to string for database storage
   */
  private normalizeMarketId(marketId: MarketId): string {
    return String(marketId);
  }

  /**
   * Store or update a market
   */
  async upsertMarket(market: MarketData): Promise<string> {
    return retryDatabaseOperation(async () => {
      try {
        const client = this.clientManager.getClient();

      // Check if market exists
      const { data: existing, error: selectError } = await client
        .from('markets')
        .select('id')
        .eq('condition_id', market.conditionId)
        .maybeSingle();

      if (selectError) {
        console.error('[DatabasePersistence] Error checking existing market:', selectError);
        throw new Error(`Failed to check existing market: ${selectError.message}`);
      }

      if (existing) {
        // Update existing market
        const updateData: TablesUpdate<'markets'> = {
          question: market.question,
          description: market.description,
          event_type: market.eventType,
          market_probability: market.marketProbability,
          volume_24h: market.volume24h,
          liquidity: market.liquidity,
          status: market.status || 'active',
          resolved_outcome: market.resolvedOutcome,
          trending_score: market.trendingScore,
          last_analyzed_at: new Date().toISOString(),
          updated_at: new Date().toISOString(),
        };

        const { error: updateError } = await client
          .from('markets')
          .update(updateData)
          .eq('id', existing.id);

        if (updateError) {
          console.error('[DatabasePersistence] Error updating market:', updateError);
          throw new Error(`Failed to update market: ${updateError.message}`);
        }

        console.log('[DatabasePersistence] Market updated successfully:', existing.id);
        return existing.id;
      } else {
        // Insert new market - use condition_id as the market ID for consistency
        const marketId = market.conditionId;
        
        const insertData: TablesInsert<'markets'> = {
          id: marketId,
          condition_id: market.conditionId,
          question: market.question,
          description: market.description,
          event_type: market.eventType,
          market_probability: market.marketProbability,
          volume_24h: market.volume24h,
          liquidity: market.liquidity,
          status: market.status || 'active',
          resolved_outcome: market.resolvedOutcome,
          trending_score: market.trendingScore,
          last_analyzed_at: new Date().toISOString(),
        };

        const { data, error: insertError } = await client
          .from('markets')
          .insert(insertData)
          .select('id')
          .single();

        if (insertError || !data) {
          console.error('[DatabasePersistence] Error inserting market:', insertError);
          throw new Error(`Failed to insert market: ${insertError?.message || 'No data returned'}`);
        }

        console.log('[DatabasePersistence] Market inserted successfully:', data.id);
        return data.id;
      }
    } catch (error) {
      console.error('[DatabasePersistence] upsertMarket failed:', error);
      throw error;
    }
    }, 'upsertMarket');
  }

  /**
   * Store a recommendation
   */
  async storeRecommendation(
    marketId: MarketId,
    recommendation: TradeRecommendation
  ): Promise<string> {
    return retryDatabaseOperation(async () => {
      try {
        const client = this.clientManager.getClient();
        const normalizedMarketId = this.normalizeMarketId(marketId);

      const insertData: TablesInsert<'recommendations'> = {
        market_id: normalizedMarketId,
        direction: recommendation.action,
        fair_probability: recommendation.metadata.consensusProbability,
        market_edge: recommendation.metadata.edge,
        expected_value: recommendation.expectedValue,
        confidence: this.mapConfidenceLevel(recommendation.metadata.confidenceBand),
        entry_zone_min: recommendation.entryZone[0],
        entry_zone_max: recommendation.entryZone[1],
        target_zone_min: recommendation.targetZone[0],
        target_zone_max: recommendation.targetZone[1],
        stop_loss: recommendation.stopLoss,
        explanation: recommendation.explanation.summary,
        core_thesis: recommendation.explanation.coreThesis,
        catalysts: recommendation.explanation.keyCatalysts,
        risks: recommendation.explanation.failureScenarios,
      };

      const { data, error } = await client
        .from('recommendations')
        .insert(insertData)
        .select('id')
        .single();

      if (error || !data) {
        console.error('[DatabasePersistence] Error storing recommendation:', error);
        throw new Error(`Failed to store recommendation: ${error?.message || 'No data returned'}`);
      }

      console.log('[DatabasePersistence] Recommendation stored successfully:', data.id);
      return data.id;
    } catch (error) {
      console.error('[DatabasePersistence] storeRecommendation failed:', error);
      throw error;
    }
    }, 'storeRecommendation');
  }

  /**
   * Store agent signals
   */
  async storeAgentSignals(
    marketId: MarketId,
    recommendationId: string,
    signals: AgentSignal[]
  ): Promise<void> {
    return retryDatabaseOperation(async () => {
      try {
        const client = this.clientManager.getClient();
        const normalizedMarketId = this.normalizeMarketId(marketId);

      const insertData: TablesInsert<'agent_signals'>[] = signals.map((signal) => ({
        market_id: normalizedMarketId,
        recommendation_id: recommendationId,
        agent_name: signal.agentName,
        agent_type: this.inferAgentType(signal.agentName),
        fair_probability: signal.fairProbability,
        confidence: signal.confidence,
        direction: signal.direction,
        key_drivers: signal.keyDrivers as Json,
        metadata: signal.metadata as Json,
      }));

      const { error } = await client.from('agent_signals').insert(insertData);

      if (error) {
        console.error('[DatabasePersistence] Error storing agent signals:', error);
        throw new Error(`Failed to store agent signals: ${error.message}`);
      }

      console.log('[DatabasePersistence] Agent signals stored successfully:', signals.length);
    } catch (error) {
      console.error('[DatabasePersistence] storeAgentSignals failed:', error);
      throw error;
    }
    }, 'storeAgentSignals');
  }

  /**
   * Record analysis history
   */
  async recordAnalysis(marketId: MarketId, analysis: AnalysisRecord): Promise<void> {
    return retryDatabaseOperation(async () => {
      try {
        const client = this.clientManager.getClient();
        const normalizedMarketId = this.normalizeMarketId(marketId);

      const insertData: TablesInsert<'analysis_history'> = {
        market_id: normalizedMarketId,
        analysis_type: analysis.type,
        status: analysis.status,
        duration_ms: analysis.durationMs,
        cost_usd: analysis.costUsd,
        agents_used: analysis.agentsUsed,
        error_message: analysis.errorMessage,
      };

      const { error } = await client.from('analysis_history').insert(insertData);

      if (error) {
        console.error('[DatabasePersistence] Error recording analysis:', error);
        throw new Error(`Failed to record analysis: ${error.message}`);
      }

      console.log('[DatabasePersistence] Analysis recorded successfully');
    } catch (error) {
      console.error('[DatabasePersistence] recordAnalysis failed:', error);
      throw error;
    }
    }, 'recordAnalysis');
  }

  /**
   * Get markets needing update
   */
  async getMarketsForUpdate(updateIntervalMs: number): Promise<MarketData[]> {
    return retryDatabaseOperation(async () => {
      try {
        const client = this.clientManager.getClient();

      // Calculate cutoff timestamp
      const cutoffTime = new Date(Date.now() - updateIntervalMs).toISOString();

      const { data, error } = await client
        .from('markets')
        .select('*')
        .eq('status', 'active')
        .or(`last_analyzed_at.is.null,last_analyzed_at.lt.${cutoffTime}`)
        .order('trending_score', { ascending: false, nullsFirst: false })
        .order('last_analyzed_at', { ascending: true, nullsFirst: true });

      if (error) {
        console.error('[DatabasePersistence] Error getting markets for update:', error);
        throw new Error(`Failed to get markets for update: ${error.message}`);
      }

      console.log('[DatabasePersistence] Found markets for update:', data?.length || 0);

      return (data || []).map((row) => ({
        conditionId: row.condition_id,
        question: row.question,
        description: row.description || undefined,
        eventType: row.event_type,
        marketProbability: row.market_probability || undefined,
        volume24h: row.volume_24h || undefined,
        liquidity: row.liquidity || undefined,
        status: row.status as 'active' | 'inactive' | 'resolved',
        resolvedOutcome: row.resolved_outcome || undefined,
        trendingScore: row.trending_score || undefined,
      }));
    } catch (error) {
      console.error('[DatabasePersistence] getMarketsForUpdate failed:', error);
      throw error;
    }
    }, 'getMarketsForUpdate');
  }

  /**
   * Mark market as resolved
   */
  async markMarketResolved(marketId: MarketId, outcome: string): Promise<void> {
    return retryDatabaseOperation(async () => {
      try {
        const client = this.clientManager.getClient();
        const normalizedMarketId = this.normalizeMarketId(marketId);

      const updateData: TablesUpdate<'markets'> = {
        status: 'resolved',
        resolved_outcome: outcome,
        updated_at: new Date().toISOString(),
      };

      const { error } = await client.from('markets').update(updateData).eq('id', normalizedMarketId);

      if (error) {
        console.error('[DatabasePersistence] Error marking market as resolved:', error);
        throw new Error(`Failed to mark market as resolved: ${error.message}`);
      }

      console.log('[DatabasePersistence] Market marked as resolved:', normalizedMarketId);
    } catch (error) {
      console.error('[DatabasePersistence] markMarketResolved failed:', error);
      throw error;
    }
    }, 'markMarketResolved');
  }

  /**
   * Get latest recommendation for a market
   */
  async getLatestRecommendation(marketId: MarketId): Promise<TradeRecommendation | null> {
    return retryDatabaseOperation(async () => {
      try {
        const client = this.clientManager.getClient();
        const normalizedMarketId = this.normalizeMarketId(marketId);

      const { data, error } = await client
        .from('recommendations')
        .select('*')
        .eq('market_id', normalizedMarketId)
        .order('created_at', { ascending: false })
        .limit(1)
        .maybeSingle();

      if (error) {
        console.error('[DatabasePersistence] Error getting latest recommendation:', error);
        throw new Error(`Failed to get latest recommendation: ${error.message}`);
      }

      if (!data) {
        console.log('[DatabasePersistence] No recommendation found for market:', normalizedMarketId);
        return null;
      }

      // Get market data for marketId
      const { data: marketData, error: marketError } = await client
        .from('markets')
        .select('condition_id')
        .eq('id', normalizedMarketId)
        .single();

      if (marketError || !marketData) {
        console.error('[DatabasePersistence] Error getting market data:', marketError);
        throw new Error(`Failed to get market data: ${marketError?.message || 'No data returned'}`);
      }

      // Reconstruct TradeRecommendation from database row
      const recommendation: TradeRecommendation = {
        marketId: marketData.condition_id,
        action: data.direction as 'LONG_YES' | 'LONG_NO' | 'NO_TRADE',
        entryZone: [data.entry_zone_min || 0, data.entry_zone_max || 0],
        targetZone: [data.target_zone_min || 0, data.target_zone_max || 0],
        stopLoss: data.stop_loss || Math.max(0.01, (data.entry_zone_min || 0) - 0.03),
        expectedValue: data.expected_value || 0,
        winProbability: data.fair_probability || 0,
        liquidityRisk: 'medium', // Default value, not stored in DB
        explanation: {
          summary: data.explanation || '',
          coreThesis: '', // Not stored separately
          keyCatalysts: (data.catalysts as string[]) || [],
          failureScenarios: (data.risks as string[]) || [],
        },
        metadata: {
          consensusProbability: data.fair_probability || 0,
          marketProbability: 0, // Not stored in recommendations table
          edge: data.market_edge || 0,
          confidenceBand: [0, 1], // Default value, not stored in DB
        },
      };

      console.log('[DatabasePersistence] Latest recommendation retrieved:', data.id);
      return recommendation;
    } catch (error) {
      console.error('[DatabasePersistence] getLatestRecommendation failed:', error);
      throw error;
    }
    }, 'getLatestRecommendation');
  }

  /**
   * Map confidence band to confidence level
   */
  private mapConfidenceLevel(confidenceBand: [number, number]): string {
    const bandWidth = confidenceBand[1] - confidenceBand[0];
    if (bandWidth < 0.1) return 'high';
    if (bandWidth < 0.2) return 'moderate';
    return 'low';
  }

  /**
   * Infer agent type from agent name
   */
  private inferAgentType(agentName: string): string {
    const name = agentName.toLowerCase();
    if (name.includes('news') || name.includes('event')) return 'event_intelligence';
    if (name.includes('polling') || name.includes('historical')) return 'polling_statistical';
    if (name.includes('sentiment') || name.includes('narrative')) return 'sentiment_narrative';
    if (name.includes('momentum') || name.includes('reversion')) return 'price_action';
    if (name.includes('catalyst') || name.includes('tail')) return 'event_scenario';
    if (name.includes('risk')) return 'risk_philosophy';
    return 'mvp';
  }
}

/**
 * Create a database persistence instance
 */
export function createDatabasePersistence(
  clientManager: SupabaseClientManager
): DatabasePersistence {
  return new DatabasePersistenceImpl(clientManager);
}
